src/BlazorWebView/src/WindowsForms/WindowsFormsDispatcher.cs

```diff
@@ -72,8 +72,28 @@ public override async Task InvokeAsync(Func<Task> workItem)
 				}
 				else
 				{
-					var asyncResult = _dispatchThreadControl.BeginInvoke(workItem);
-					await Task.Factory.FromAsync(asyncResult, _dispatchThreadControl.EndInvoke);
+					// See https://github.com/dotnet/winforms/issues/4631 for discussion. `Control.BeginInvoke` in WinForms
+					// does not wait for Tasks returned by the delegate. We will have to simulate this using a TCS and wait for
+					// both execution of `workItem` and the dispatcher to complete its internal operation.
+					// additional APIs are exposed by WinForms.
+
+					var tcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+					// BeginInvoke specifically expects an `Action` so avoid using var here.
+					Action action = async () =>
+					{
+						try
+						{
+							await workItem();
+							tcs.TrySetResult();
+						}
+						catch (Exception ex)
+						{
+							tcs.TrySetException(ex);
+						}
+					};
+
+					var asyncResult = _dispatchThreadControl.BeginInvoke(action, workItem, tcs);
+					await Task.WhenAll(tcs.Task, Task.Factory.FromAsync(asyncResult, _dispatchThreadControl.EndInvoke));
```

Comment:
Sorry, I didn't have chance to review this before (as the author of https://github.com/dotnet/maui/issues/2313).

I believe you only need `await tcs.Task` here. 

awaiting `Task.Factory.FromAsync(asyncResult, _dispatchThreadControl.EndInvoke)` in parallel with it is at best redundant. 

cc @Eilon @pranavkm

Reply:
I wanted to make sure we were awaiting the dispatch in the event there were operations it needed to wrap up after the delegate was executed. By the way, in the case that we're already in the right context, these functions execute inline without this additional overhead. So in the ordinary case, you only pay this extra cost if you're calling from a background thread.